src/Xamarin.Android.Build.Tasks/Tasks/GetAdditionalResourcesFromAssemblies.cs

```diff
@@ -193,6 +194,71 @@ bool IsValidDownload(string file, string sha1)
 			return string.Compare (hash, sha1, StringComparison.InvariantCultureIgnoreCase) == 0;
 		}
 
+		void DoDownload (long totalBytes, long offset, Stream responseStream, Stream outputStream, Action<long, long, int> progressCallback = null)
+		{
+			long readSoFar = offset;
+			byte [] buffer = new byte [8192];
+			int bufferSize = buffer.Length;
+			int nread = 0;
+			float percent, tb = (float)totalBytes, lastPercent = 0;
+			outputStream.Seek (offset, SeekOrigin.Begin);
+
+			do {
+				nread = responseStream.Read (buffer, 0, bufferSize);
+				if (nread <= 0)
+					break;
+				readSoFar += nread;
+				outputStream.Write (buffer, 0, nread);
+				percent = (float)readSoFar / tb * 100;
+				if (percent - lastPercent > 1) {
+					progressCallback?.Invoke (readSoFar, totalBytes, (int)percent);
+					lastPercent = percent;
+				}
+				outputStream.Flush ();
+			} while (true);
+		}
+
+		void Download (string file, Uri uri, Action<long, long, int> progressCallback = null)
+		{
+			var request = WebRequest.CreateHttp (uri);
+			int offset = 0;
+			if (File.Exists (file) && !MonoAndroidHelper.IsValidZip (file)) {
+				var fi = new FileInfo (file);
+				request.AddRange (fi.Length);
+				offset = (int)fi.Length;
+				LogMessage ("Partial download detected. Resuming from previous download progress.");
+			}
+			if (!File.Exists (file) || offset > 0) {
+				HttpWebResponse response = null;
+				try {
+					response = (HttpWebResponse)request.GetResponse ();
+				}
+				catch (WebException ex) {
+					var exceptionResponse = ex.Response as HttpWebResponse;
+					if (exceptionResponse?.StatusCode == HttpStatusCode.RequestedRangeNotSatisfiable) {
+						// Download the entire file again.
+						request.Abort ();
+						request = WebRequest.CreateHttp (uri);
+						File.Delete (file);
+						offset = 0;
+						request.AddRange (0);
+						response = (HttpWebResponse)request.GetResponse ();
+						LogMessage ("Could not resume previous download. Starting again.");
+					}
+					else
+						throw;
```

Comment:
Can we "invert" this logic to avoid the extra nesting/promote the `throw`?

    if (exceptionResponse?.StatusCode != HttpStatusCode.RequestedRangeNotSatisfiable)
        throw;
    ...